modyn/supervisor/internal/triggers/datadrifttrigger.py

```diff
@@ -127,41 +154,85 @@ def inform(
-
-        if offset + len(new_key_ts) < detect_interval:
-            # No detection in this trigger
-            self._update_curr_window(new_key_ts)
-            return
-
-        # At least one detection, fill up window up to that detection
-        self._update_curr_window(new_key_ts[: detect_interval - offset])
-        new_key_ts = new_key_ts[detect_interval - offset :]
-        trigger_idx = detect_interval - offset - 1  # If we trigger, it will be on this index
-
-        if not self._triggered_once:
-            # If we've never triggered before, always trigger
-            self._triggered_once = True
-            triggered = True
-            drift_results: dict[str, MetricResult] = {}
-        else:
-            # Run the detection
-            triggered, drift_results = self._run_detection()
-
-        yield from self._handle_drift_result(triggered, trigger_idx, drift_results, log=log)
+
+        # index of the first unprocessed data point in the batch
+        processing_head_in_batch = 0
 
         # Go through remaining data in new data in batches of `detect_interval`
-        for i in range(0, len(new_key_ts), detect_interval):
-            batch = new_key_ts[i : i + detect_interval]
-            trigger_idx += detect_interval
-            self._update_curr_window(batch)
-
-            if len(batch) == detect_interval:
-                # Regular batch, in this case run detection
-                triggered, drift_results = self._run_detection()
+        while True:
+            if self._sample_left_until_detection - len(new_key_ts) > 0:
+                # No detection in this trigger because of too few data points to fill detection interval
+                self._windows.inform_data(new_key_ts)  # update current window
+                self._sample_left_until_detection -= len(new_key_ts)
+                return
+
+            # At least one detection, fill up window up to that detection
+            next_detection_interval = new_key_ts[: self._sample_left_until_detection]
+            self._windows.inform_data(next_detection_interval)
+
+            # Update the remaining data
+            processing_head_in_batch += len(next_detection_interval)
+            new_key_ts = new_key_ts[len(next_detection_interval) :]
+
+            # Reset for next detection
+            self._sample_left_until_detection = self.config.detection_interval_data_points
+
+            # The first ever detection will always trigger
+            if not self._triggered_once or len(self.warmup_intervals) < (self.config.warmup_intervals or 0):
+                # If we've never triggered before, always trigger
+                self._triggered_once = True
+                triggered = True
+                drift_results: dict[str, MetricResult] = {}
+                if len(self.warmup_intervals) < (self.config.warmup_intervals or 0):
+                    self.warmup_intervals.append(list(self._windows.reference))
```

Comment:
it seems we currently don't have a test for the warmup case (according to codecov) (but I am not too sure about the list of intervals in the first case. in my understanding so far warmup is either True or False)

Reply:
I added a quite extensive test, good that you mentioned this, there were still a couple of bugs in the implementation.